quint/src/types/constraintGenerator.ts

```diff
@@ -152,31 +177,47 @@ export class ConstraintGeneratorVisitor implements IRVisitor {
       })
     )
 
-    const result = argsResult.chain((results): Either<Error, TypeScheme> => {
-      const signature = this.typeForName(e.opcode, e.id, e.args.length)
-      const a: QuintType = { kind: 'var', name: this.freshVarGenerator.freshVar('t') }
-      const special = specialConstraints(e.opcode, e.id, results, a)
-
-      const constraints = special.chain(cs => {
-        // Check if there is a special case defined for the operator
-        if (cs.length > 0) {
-          // If yes, use the special constraints
-          return right(cs)
-        } else {
-          // Otherwise, define a constraint over the signature
-          return signature.map(t1 => {
-            const t2: QuintType = { kind: 'oper', args: results.map(r => r[1]), res: a }
-            const c: Constraint = { kind: 'eq', types: [t1, t2], sourceId: e.id }
-            return [c]
-          })
+    // We want `definedSignature` computed before the fresh variable `a` so that the
+    // numbering of ther fresh variables stays in order, with `a`, used for return types,
+    // bearing the highest number.
```

Comment:
Thanks for adding that! I realize I should have done that when I faced this issue :sweat_smile: